src/ReferenceAnalyzer.Core/ReferenceAnalyzer.cs

```diff
@@ -181,22 +203,38 @@ private async Task<Compilation> GetCompilation(Project project, string outputPat
             return compilation;
         }
 
-        public async IAsyncEnumerable<ReferencesReport> AnalyzeAll()
+        public IAsyncEnumerable<ReferencesReport> AnalyzeAll()
+        {
+            return AnalyzeAll(CancellationToken.None);
+        }
+
+        public async IAsyncEnumerable<ReferencesReport> AnalyzeAll([EnumeratorCancellation] CancellationToken token)
         {
-            await foreach (var report in Analyze(_projects.Select(p => p.Name)))
+            await foreach (var report in Analyze(_projects.Select(p => p.Name), token))
                 yield return report;
         }
 
-        public async IAsyncEnumerable<ReferencesReport> Analyze(IEnumerable<string> projects)
+        public IAsyncEnumerable<ReferencesReport> Analyze(IEnumerable<string> projects)
+        {
+            return Analyze(projects, CancellationToken.None);
+        }
+
+        public async IAsyncEnumerable<ReferencesReport> Analyze(IEnumerable<string> projects,
+            [EnumeratorCancellation] CancellationToken token)
         {
+            if (projects == null)
+                throw new ArgumentNullException(nameof(projects));
+
+            token.Register(() => ProgressReporter.Report(0));
+
             var analyzedProjects = 0;
             var totalProjects = projects.Count();
-            ProgressReporter.Report(0);
+            ProgressReporter.Report(-1);
```

Comment:
Maybe it is time to add some const for Report() arguments. I had to think for a minute what does -1 mean in this context.